Rust Implementation

clarkmoody opened the pull request:
Bech32 encoding of SegWit addresses in a Rust Crate.

Instructions are in README

sipa commented:
Cool, is this ready? I'll find someone to review the code then (unfortunately I do not know Rust yet).

clarkmoody commented:
I may change the API, so I'll update the PR when it's ready for review.
Needs more docs too :-)

On Sat, Apr 1, 2017, 11:19 Pieter Wuille <notifications@github.com> wrote:

> Cool, is this ready? I'll find someone to review the code then
> (unfortunately I do not know Rust yet).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sipa/bech32/pull/12#issuecomment-290930198>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABfXMdpwgA3dfqFpWauQRczusykHMdMsks5rrnkEgaJpZM4Mtpv->
> .
>

clarkmoody commented:
The Rust library is ready for review.

apoelstra commented:
Hi, I can take a look at this.

apoelstra commented:
Can you squash the first three commits and the wit_prof rename into one commit?

clarkmoody commented:
I don't think I can put the license data commit (a55881c) after the rename commit (f086b45).

How a squash to three commits total?

1. First three commits: Initial implementation and tests passing
2. Add license and package data
3. Last three commits: Rename + refactor API into final version

apoelstra commented:
Sounds good to me!

clarkmoody commented:
Commits pushed.

apoelstra commented:
On a plane in about 8 hours, will be out of commission for a day or two while I adjust to the timezone. But I haven't forgotten about this :)

apoelstra commented:
ACK except for

- various nits I listed (though having no global `Error` type is fine)
- scan the code to delete any trailing spaces from lines

Otherwise this looks great!

clarkmoody commented:
@apoelstra Did you make comments beyond commit squash above? Those may have been lost.

I've brought my fork up to `sipa:master` in the meantime, removed a few trailing spaces, and added a simple README.

apoelstra commented:
Can you see my comments now? I didn't realize github had added several extra button clicks to comment.

apoelstra commented on ref/rust/src/lib.rs:
```diff
@@ -0,0 +1,264 @@
+pub mod bech32;
+pub mod wit_prog;
+
+/// Error types for Bech32 encoding / decoding
```
I wouldn't worry about this one. It's a lot of boilerplate code that would distract from the real stuff. If this were an independent crate I'd say it should exist.

clarkmoody replied:
I would like to publish as a crate later. My fork may diverge at that point to add the extras to make it happen.

clarkmoody commented:
Thanks, I see them now. Looks like the comments were for a previous commit of the code, but most of them are still applicable.

clarkmoody commented on ref/rust/src/segwit_addr.rs:
```diff
@@ -1,7 +1,174 @@
+pub fn encode(hrp: String, p: WitnessProgram) -> EncodeResult {
+    let mut data: Vec<u8> = vec![p.version];
+    // Convert 8-bit program into 5-bit
+    let p5 = match convert_bits(p.program.to_vec(), 8, 5, true) {
```
Since `convert_bits` is used for both encoding and decoding, it returns a result type containing a `CodingError`.

Elsewhere you mentioned a single `Error` type. I may have gone overboard in defining lots of errors, but the error hierarchy may be useful. Each function may return each `enum` value for its error type and no more or less. The alternative would be a flat `Error` struct with all error values in it. This would enable code shortcuts like `try!` and `?`. Would you recommend restructuring errors to be a flat `enum`?

clarkmoody commented on ref/rust/src/segwit_addr.rs:
```diff
@@ -1,7 +1,174 @@
+        if v > 0x80 {
+            v -= 0x80;
+        }
+        let (_, program) = pubkey.split_at(2);
```
Added in latest commit.

clarkmoody commented on ref/rust/src/bech32.rs:
```diff
@@ -114,6 +120,19 @@ pub fn decode(s: String) -> DecodeResult {
+    let mut values: Vec<u8> = hrp_expand(hrp);
+    values.extend_from_slice(data);
+    let pad: Vec<u8> = vec![0,0,0,0,0,0];
+    values.extend_from_slice(&pad);
```
Added in latest commit.

clarkmoody commented on ref/rust/src/bech32.rs:
```diff
@@ -1,12 +1,24 @@
+            data: self.data.to_vec()
+        }
+    }
 }
```
I need to be more familiar with the various things you can `#derive`!

Added in latest commit.

clarkmoody commented on ref/rust/src/bech32.rs:
```diff
@@ -0,0 +1,190 @@
+        }
+        let mut c = b;
+        // Lowercase
+        if b >= 97 && b <= 122 {
```
Good call. This makes it so much easier to read the code than ASCII magic in there.

Added in latest commit.

clarkmoody commented on ref/rust/src/segwit_addr.rs:
```diff
@@ -1,7 +1,174 @@
+        pubkey
+    }
+
+    pub fn from_scriptpubkey(pubkey: Vec<u8>) -> Result<WitnessProgram, ()> {
```
Added in latest commit along with tests in the valid address battery.

apoelstra commented:
ACK.

@sipa do you have any opinion on whether these six commits should be squashed into one?

clarkmoody commented:
Merged in changes from `sipa:master`

sipa commented:
Yes, I prefer a clean commit history. No need to squash everything into one if there are logically separable parts, but no merges if possible.

clarkmoody commented:
Well it looks like I made a mess of the commits by merging in `sipa:master` changes along the way. What should I do now?

sipa commented:
I can squash them for you if you want.

clarkmoody commented:
That would be great, thanks :smile:

sipa commented:
@clarkmoody See #15.

sipa commented:
Superseded by #15.
